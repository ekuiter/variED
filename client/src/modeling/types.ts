import {HierarchyPointNode} from 'd3-hierarchy';
import {Point} from '../types';

// tags and attributes used in serialized feature models
export const STRUCT = 'struct';
export const CONSTRAINTS = 'constraints';
export const PROPERTIES = 'properties';
export const CALCULATIONS = 'calculations';
export const COMMENTS = 'comments';
export const FEATURE_ORDER = 'featureOrder';
export const TYPE = 'type';
export const NAME = 'name';
export const UUID = 'uuid';
export const DESCRIPTION = 'description';
export const MANDATORY = 'mandatory';
export const ABSTRACT = 'abstract';
export const HIDDEN = 'hidden';
export const VAR = 'var';

export enum FeatureType {
    feature = 'feature',
    or = 'or',
    alt = 'alt',
    and = 'and',
    unknown = 'unknown'
};

enum ConstraintType {
    var = 'var',
    not = 'not',
    disj = 'disj',
    eq = 'eq',
    imp = 'imp',
    conj = 'conj',
    atmost1 = 'atmost1',
    unknown = 'unknown'
};

export interface SerializedFeatureNode {
    [TYPE]: FeatureType,
    [NAME]: string,
    [UUID]: string,
    [HIDDEN]?: boolean,
    [MANDATORY]?: boolean,
    [ABSTRACT]?: boolean,
    [DESCRIPTION]?: string,
    children?: SerializedFeatureNode[]
};

interface SerializedConstraintNode {
    [TYPE]: ConstraintType,
    [VAR]?: string,
    children?: SerializedConstraintNode[]
};

export interface SerializedConstraint {
    [TYPE]: 'rule',
    children: SerializedConstraintNode[]
};

export interface SerializedFeatureModel {
    [STRUCT]: SerializedFeatureNode[],
    [CONSTRAINTS]: SerializedConstraint[],
    // ignored for now
    [CALCULATIONS]: never,
    [COMMENTS]: never,
    [FEATURE_ORDER]: never
};

export type FeaturePropertyKey = string | ((node: GraphicalFeatureNode) => string);

export interface GraphicalFeature {
    node: GraphicalFeatureNode,
    uuid: string,
    name: string,
    type: FeatureType,
    description?: string,
    isRoot: boolean,
    isAbstract: boolean,
    isHidden: boolean,
    isMandatory: boolean,
    isAnd: boolean,
    isOr: boolean,
    isAlternative: boolean,
    isGroup: boolean,
    isCollapsed: boolean,
    hasChildren: boolean,
    hasActualChildren: boolean,
    getPropertyString: (key: FeaturePropertyKey) => string,
    getNumberOfFeaturesBelow: () => number
};

type Datum = SerializedFeatureNode; // accessible as node.data

export type GraphicalFeatureNode = HierarchyPointNode<Datum> & {
    children?: GraphicalFeatureNode[];
    actualChildren?: GraphicalFeatureNode[];
    _feature: GraphicalFeature;
    feature: () => GraphicalFeature;
}

export type NodeCoordinateFunction = (node: GraphicalFeatureNode) => number;
export type NodeCoordinateForAxisFunction = (node: GraphicalFeatureNode, axis: string) => number;
export type NodePointFunction = (node: GraphicalFeatureNode) => Point;