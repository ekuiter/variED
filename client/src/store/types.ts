import FeatureModel from '../modeling/FeatureModel';
import {defaultSettings, Settings} from './settings';
import {Message, FeatureDiagramLayoutType, OverlayType, OverlayProps, ArtifactPath} from '../types';
import {Feature, SerializedFeatureModel} from '../modeling/types';

export interface Collaborator {
    siteID: string
};

export interface CollaborativeSession {
    artifactPath: ArtifactPath,
    collaborators: Collaborator[]
};

export type KernelContext = object;

export interface FeatureDiagramCollaborativeSession extends CollaborativeSession {
    kernelContext: KernelContext,
    serializedFeatureModel: SerializedFeatureModel,
    layout: FeatureDiagramLayoutType,
    isSelectMultipleFeatures: boolean,
    selectedFeatureIDs: string[],
    collapsedFeatureIDs: string[]
};

export interface State {
    settings: Settings,
    overlay: OverlayType,
    overlayProps: OverlayProps
    myself?: Collaborator,
    collaborativeSessions: CollaborativeSession[],
    artifactPaths: ArtifactPath[],
    currentArtifactPath?: ArtifactPath
};

export const initialState: State = {
    settings: defaultSettings,
    overlay: OverlayType.none,
    overlayProps: {},
    myself: undefined,
    collaborativeSessions: [],
    artifactPaths: [],
    currentArtifactPath: undefined
};

export const initialFeatureDiagramCollaborativeSessionState =
    (artifactPath: ArtifactPath, kernelContext: KernelContext, serializedFeatureModel: SerializedFeatureModel):
    FeatureDiagramCollaborativeSession => ({
        artifactPath,
        collaborators: [],
        kernelContext,
        serializedFeatureModel,
        layout: FeatureDiagramLayoutType.verticalTree,
        isSelectMultipleFeatures: false,
        selectedFeatureIDs: [],
        collapsedFeatureIDs: []
    });

export type OnSelectFeatureFunction = (payload: {featureID: string}) => void;
export type OnDeselectFeatureFunction = (payload: {featureID: string}) => void;
export type OnSelectAllFeaturesFunction = () => void;
export type OnDeselectAllFeaturesFunction = () => void;
export type OnCollapseAllFeaturesFunction = () => void;
export type OnExpandAllFeaturesFunction = () => void;
export type OnSetFeatureDiagramLayoutFunction = (payload: {layout: FeatureDiagramLayoutType}) => void;
export type OnSetSelectMultipleFeaturesFunction = (payload: {isSelectMultipleFeatures: boolean}) => void;
export type OnCollapseFeaturesFunction = (payload: {featureIDs: string[]}) => void;
export type OnExpandFeaturesFunction = (payload: {featureIDs: string[]}) => void;
export type OnCollapseFeaturesBelowFunction = (payload: {featureIDs: string[]}) => void;
export type OnExpandFeaturesBelowFunction = (payload: {featureIDs: string[]}) => void;
export type OnShowOverlayFunction = (payload: {overlay: OverlayType, overlayProps: OverlayProps, selectOneFeatureID?: string}) => void;
export type OnHideOverlayFunction = (payload: {overlay: OverlayType}) => void;
export type OnFitToScreenFunction = () => void;
export type OnSetSettingFunction = (payload: {path: string, value: any}) => void;
export type OnResetSettingsFunction = () => void;
export type OnSetCurrentArtifactPathFunction = (payload: {artifactPath?: ArtifactPath}) => void;

export type OnJoinRequestFunction = (payload: {artifactPath: ArtifactPath}) => Promise<void>;
export type OnLeaveRequestFunction = (payload: {artifactPath: ArtifactPath}) => Promise<void>;
export type OnUndoFunction = () => Promise<void>;
export type OnRedoFunction = () => Promise<void>;
export type OnCreateFeatureBelowFunction = (payload: {featureParentID: string}) => Promise<void>;
export type OnCreateFeatureAboveFunction = (payload: {featureIDs: string[]}) => Promise<void>;
export type OnRemoveFeatureFunction = (payload: {featureIDs: string[]}) => Promise<void>;
export type OnRemoveFeatureSubtreeFunction = (payload: {featureIDs: string[]}) => Promise<void>;
export type OnSetFeatureNameFunction = (payload: {featureID: string, name: string}) => Promise<void>;
export type OnSetFeatureDescriptionFunction = (payload: {featureID: string, description: string}) => Promise<void>;
export type OnSetFeatureAbstractFunction = (payload: {featureIDs: string[], value: boolean}) => Promise<void>;
export type OnSetFeatureHiddenFunction = (payload: {featureIDs: string[], value: boolean}) => Promise<void>;
export type OnSetFeatureOptionalFunction = (payload: {featureIDs: string[], value: boolean}) => Promise<void>;
export type OnToggleFeatureOptionalFunction = (payload: {feature: Feature}) => Promise<void>;
export type OnSetFeatureAndFunction = (payload: {featureIDs: string[]}) => Promise<void>;
export type OnSetFeatureOrFunction = (payload: {featureIDs: string[]}) => Promise<void>;
export type OnSetFeatureAlternativeFunction = (payload: {featureIDs: string[]}) => Promise<void>;
export type OnToggleFeatureGroupTypeFunction = (payload: {feature: Feature}) => Promise<void>;

// Props that may derived from the state to use in React components.
// This enforces the convention that a prop called 'on...' has the same type in all components.
export type StateDerivedProps = Partial<{
    handleMessage: (message: Message) => void,
    currentArtifactPath: ArtifactPath,
    artifactPaths: ArtifactPath[],
    collaborativeSessions: CollaborativeSession[],
    myself: Collaborator,
    collaborators: Collaborator[],
    settings: Settings,
    featureDiagramLayout: FeatureDiagramLayoutType,
    isSelectMultipleFeatures: boolean,
    selectedFeatureIDs: string[],
    featureModel: FeatureModel,
    overlay: OverlayType,
    overlayProps: OverlayProps,

    onSelectFeature: OnSelectFeatureFunction,
    onDeselectFeature: OnDeselectFeatureFunction,
    onSelectAllFeatures: OnSelectAllFeaturesFunction,
    onDeselectAllFeatures: OnDeselectAllFeaturesFunction,
    onCollapseAllFeatures: OnCollapseAllFeaturesFunction,
    onExpandAllFeatures: OnExpandAllFeaturesFunction,
    onSetFeatureDiagramLayout: OnSetFeatureDiagramLayoutFunction,
    onSetSelectMultipleFeatures: OnSetSelectMultipleFeaturesFunction,
    onCollapseFeatures: OnCollapseFeaturesFunction,
    onExpandFeatures: OnExpandFeaturesFunction,
    onCollapseFeaturesBelow: OnCollapseFeaturesBelowFunction,
    onExpandFeaturesBelow: OnExpandFeaturesBelowFunction,
    onShowOverlay: OnShowOverlayFunction,
    onHideOverlay: OnHideOverlayFunction,
    onFitToScreen: OnFitToScreenFunction,
    onSetSetting: OnSetSettingFunction,
    onResetSettings: OnResetSettingsFunction,
    onSetCurrentArtifactPath: OnSetCurrentArtifactPathFunction,

    onJoinRequest: OnJoinRequestFunction,
    onLeaveRequest: OnLeaveRequestFunction,
    onUndo: OnUndoFunction,
    onRedo: OnRedoFunction,
    onCreateFeatureBelow: OnCreateFeatureBelowFunction,
    onCreateFeatureAbove: OnCreateFeatureAboveFunction,
    onRemoveFeature: OnRemoveFeatureFunction,
    onRemoveFeatureSubtree: OnRemoveFeatureSubtreeFunction,
    onSetFeatureName: OnSetFeatureNameFunction,
    onSetFeatureDescription: OnSetFeatureDescriptionFunction,
    onSetFeatureAbstract: OnSetFeatureAbstractFunction,
    onSetFeatureHidden: OnSetFeatureHiddenFunction,
    onSetFeatureOptional: OnSetFeatureOptionalFunction,
    onToggleFeatureOptional: OnToggleFeatureOptionalFunction,
    onSetFeatureAnd: OnSetFeatureAndFunction,
    onSetFeatureOr: OnSetFeatureOrFunction,
    onSetFeatureAlternative: OnSetFeatureAlternativeFunction,
    onToggleFeatureGroupType: OnToggleFeatureGroupTypeFunction
}>;