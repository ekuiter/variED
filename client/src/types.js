import PropTypes from 'prop-types';

export const
    layoutTypes = {
        verticalTree: 'verticalTree',
        horizontalTree: 'horizontalTree'
    },
    LayoutType = PropTypes.oneOf(Object.values(layoutTypes)),

    formatTypes = {
        svg: 'svg',
        png: 'png',
        jpg: 'jpg',
        pdf: 'pdf'
    },
    FormatType = PropTypes.oneOf(Object.values(formatTypes)),

    SettingsType = PropTypes.object,

    overlayTypes = {
        settingsPanel: 'settingsPanel',
        aboutPanel: 'aboutPanel',
        featurePanel: 'featurePanel',
        featureRenameDialog: 'featureRenameDialog',
        featureSetDescriptionDialog: 'featureSetDescriptionDialog',
        exportDialog: 'export',
        featureCallout: 'featureCallout',
        featureContextualMenu: 'featureContextualMenu',
        isFloatingFeature: type => type === 'featureCallout' || type === 'featureContextualMenu'
    },
    OverlayType = PropTypes.oneOf(Object.values(overlayTypes));