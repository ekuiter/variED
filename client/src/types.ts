/**
 * Definition of common types and enumerations.
 */

import {Selection} from 'd3-selection';

export enum MessageType {
    ERROR = 'ERROR',
    COLLABORATOR_INFO = 'COLLABORATOR_INFO',
    ADD_ARTIFACT = 'ADD_ARTIFACT',
    REMOVE_ARTIFACT = 'REMOVE_ARTIFACT',
    JOIN_REQUEST = 'JOIN_REQUEST',
    LEAVE_REQUEST = 'LEAVE_REQUEST',
    INITIALIZE = 'INITIALIZE',
    KERNEL = 'KERNEL'
};

export enum FeatureDiagramLayoutType {
    verticalTree = 'verticalTree',
    horizontalTree = 'horizontalTree'
};

export enum OverlayType {
    none = 'none',
    commandPalette = 'commandPalette',
    settingsPanel = 'settingsPanel',
    aboutPanel = 'aboutPanel',
    featurePanel = 'featurePanel',
    featureRenameDialog = 'featureRenameDialog',
    featureSetDescriptionDialog = 'featureSetDescriptionDialog',
    addArtifactPanel = 'addArtifact',
    shareDialog = 'share',
    exportDialog = 'export',
    featureCallout = 'featureCallout',
    featureContextualMenu = 'featureContextualMenu'
};

export interface OverlayProps {
    featureID?: string,
    format?: FormatType
};

export function isFloatingFeatureOverlay(type: OverlayType): boolean {
    return type === OverlayType.featureCallout || type === OverlayType.featureContextualMenu;
}

export enum FormatType {
    svg = 'svg',
    png = 'png',
    jpg = 'jpg',
    pdf = 'pdf'
};

export interface FormatOptions {
    scale?: number,
    quality?: number
};

export interface ArtifactPath {
    project: string,
    artifact: string
};

export function artifactPathToString(artifactPath: ArtifactPath): string {
    return artifactPath ? `${artifactPath.project}/${artifactPath.artifact}` : `(unknown artifact)`;
}

export function isArtifactPathEqual(a?: ArtifactPath, b?: ArtifactPath): boolean {
    return typeof a !== 'undefined' && typeof b !== 'undefined' &&
        a.project === b.project && a.artifact === b.artifact;
}

export interface Message {
    type: MessageType,
    artifactPath?: ArtifactPath,
    [x: string]: any
};

export function isMessageType(type: string) {
    return Object.values(MessageType).includes(type);
};

export type Point = {x: number, y: number};
// essentially the same as a DOMRect/ClientRect
export type Rect = Point & {width: number, height: number};
// upper left and lower right point, can be transformed to rect (but the tuple form is also useful)
export type Bbox = [[number, number], [number, number]];
// Represents any selection. We ignore the generic parameters here for simplicity.
// (IMO adding them would improve type safety, but decrease overall readability.)
export type D3Selection = Selection<any, any, any, any>;
// matches any function - this should be used sparely
export type Func = (...args: any[]) => any;
// omits a property from a type
export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;