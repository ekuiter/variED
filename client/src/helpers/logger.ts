enum LogLevel {none, warn, log, info};
const defaultLogLevel = LogLevel.log;
let logLevel = defaultLogLevel;
type LogEntry = () => any;
type ConsoleFunction = (...args: any[]) => void;

const mapStateToPropsCache = {},
    
    truncate = (string: string, length = 80): string =>
        string.substr(0, length - 1) + (string.length > length ? 'â€¦' : ''),
    
    shouldSerialize = (value: any): boolean =>
        value && (
            value !== Object(value) || // may serialize primitive types
            value.constructor.name === 'Object' || // may serialize simple objects (no class instances)
            // may serialize array (if its elements should also be serialized)
            (Array.isArray(value) && value.reduce((acc, val) => acc && shouldSerialize(val), true))),

    stringify = (value: any): string =>
        // stringifies a value by serializing it or calling .toString() for complex (e.g., circular) objects
        truncate(shouldSerialize(value) ? JSON.stringify(value) : String(value)),

    taggedWrapper = (consoleFn: ConsoleFunction, minimumLogLevel: LogLevel) => ({tag, color = 'white', backgroundColor = 'slategrey'}:
        {tag: string, color?: string, backgroundColor?: string}, logEntry: LogEntry): void => {
        if (logLevel >= minimumLogLevel && logEntry)
            consoleFn(`%c${tag.toUpperCase()}`,
                `color: ${color}; background-color: ${backgroundColor}; padding: 2px 5px; font-weight: bold`,
                logEntry());
    },

    coloredWrapper = (consoleFn: ConsoleFunction, minimumLogLevel: LogLevel) => ({color = 'inherit', backgroundColor = 'inherit'}:
        {color?: string, backgroundColor?: string}, logEntry: LogEntry): void => {
        if (logLevel >= minimumLogLevel && logEntry)
            consoleFn(`%c${logEntry()}`, `color: ${color}; background-color: ${backgroundColor}; padding: 1px 3px`);
    },

    wrapper = (consoleFn: ConsoleFunction, minimumLogLevel: LogLevel) => (logEntry?: LogEntry): void => {
        if (logLevel >= minimumLogLevel)
            consoleFn(logEntry ? logEntry() : undefined);
    },

    consoleWarn = console.warn.bind(console),
    consoleLog = console.log.bind(console),
    consoleInfo = console.info.bind(console);

const logger = {
    warn: wrapper(consoleWarn, LogLevel.warn),
    log: wrapper(consoleLog, LogLevel.log),
    info: wrapper(consoleInfo, LogLevel.info),
    infoBegin: wrapper(console.group.bind(console), LogLevel.info),
    infoBeginCollapsed: wrapper(console.groupCollapsed.bind(console), LogLevel.info),
    infoEnd: wrapper(console.groupEnd.bind(console), LogLevel.info),
    warnTagged: taggedWrapper(consoleWarn, LogLevel.warn),
    logTagged: taggedWrapper(consoleLog, LogLevel.log),
    infoTagged: taggedWrapper(consoleInfo, LogLevel.info),
    warnColored: coloredWrapper(consoleWarn, LogLevel.warn),
    logColored: coloredWrapper(consoleLog, LogLevel.log),
    infoColored: coloredWrapper(consoleInfo, LogLevel.info),

    mapStateToProps(containerName: string, mapStateToProps: (state: object) => object) {
        return (state: object): object => {
            const newProps = mapStateToProps(state);
            if (logLevel >= LogLevel.info && mapStateToPropsCache[containerName] &&
                Object.entries(newProps).some(([key, value]) => value !== mapStateToPropsCache[containerName][key])) {
                logger.infoBeginCollapsed(() => `[mapStateToProps] props changed for ${containerName}`);
                Object.entries(newProps).forEach(([key, value]) => {
                    const oldValue = mapStateToPropsCache[containerName][key];
                    if (value !== oldValue) {
                        logger.infoBeginCollapsed(() => key);
                        logger.infoColored({color: 'black', backgroundColor: '#faa'}, () => stringify(oldValue));
                        logger.infoColored({color: 'black', backgroundColor: '#afa'}, () => stringify(value));
                        logger.infoEnd();
                    }
                });
                logger.infoEnd();
            }
            return mapStateToPropsCache[containerName] = newProps;
        };
    }
};

declare var window: any;
window.app = window.app || {};
window.app.setLogLevel = (_logLevel = defaultLogLevel) => logLevel = _logLevel;
window.app.LogLevel = LogLevel;

export default logger;