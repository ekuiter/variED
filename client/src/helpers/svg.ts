/**
 * Helpers for drawing SVG elements.
 * Depends on the D3.js library to manipulate elements.
 */

import {Point, D3Selection, Func} from '../types';
import {ValueMap} from 'd3-selection-multi';
import logger from './logger';
import {FeaturePropertyKey, NodeCoordinateFunction, FeatureNode, NodePointFunction} from '../modeling/types';

export type Style = ValueMap<any, any>;
export interface StyleDescriptor {
    property?: FeaturePropertyKey,
    [x: string]: any
};
export type ArcPathFunction = (center: Point, radius: number, startAngle: number, endAngle: number, sweepFlag?: boolean) => string;

function toPath(point: Point): string;
function toPath(x: number, y: number): string;
function toPath(x: number | {x: number, y: number}, y?: number): string {
    if (typeof x === 'object')
        ({x, y} = x);
    return `${x},${y}`;
}

const MOVE = 'M', LINE = 'L', CURVE = 'C', ARC = 'A', CLOSE = 'Z';

function toD(...data: (string | number)[]): string {
    return data.join(' ');
}

export function attrIfPresent(selection: D3Selection, key: string, value?: any): void {
    if (typeof value !== 'undefined')
        selection.attr(key, value);
}

export function styleIfPresent(selection: D3Selection, style?: Style): void {
    if (typeof style !== 'undefined')
        selection.call(addStyle, style);
}

export function fnIfPresent(selection: D3Selection, fn?: Func): void {
    if (typeof fn !== 'undefined')
        selection.call(fn);
}

export function appendCross(selection: D3Selection, style?: Style): void {
    style = style || {
        fill: 'none',
        stroke: 'red',
        'stroke-width': '1.5px'
    };
    selection.append('path').attr('d', toD(MOVE, -4, -4, LINE, 4, 4)).attrs(style);
    selection.append('path').attr('d', toD(MOVE, -4, 4, LINE, 4, -4)).attrs(style);
}

export function updateRect(selection: D3Selection, {x, y, klass, width, height, style, fn}:
    {x: number, y: number, klass?: string, width: number, height: number, style?: Style, fn?: Func}): void {
    style = style || {
        fill: 'none',
        stroke: 'red',
        'stroke-dasharray': '4,4'
    };
    selection
        .call(attrIfPresent, 'class', klass)
        .call(fnIfPresent, fn)
        .call(styleIfPresent, style)
        .attr('width', width)
        .attr('height', height)
        .attr('x', x)
        .attr('y', y);
}

export function translateTransform(selection: D3Selection, x: NodeCoordinateFunction, y: NodeCoordinateFunction): void {
    selection.attr('transform', (d: FeatureNode) => `translate(${toPath(x(d), y(d))})`);
}

export function drawLine(selection: D3Selection, selector: string | undefined,
    {klass, from, to, style, fn}:
    {klass?: string, from: NodePointFunction, to: NodePointFunction, style?: Style, fn?: Func}): void {
    (!selector ? selection.append('path') : selection.select(selector))
        .call(attrIfPresent, 'class', klass)
        .call(styleIfPresent, style)
        .call(fnIfPresent, fn)
        .attr('d', (d: FeatureNode) => toD(MOVE, toPath(from(d)), LINE, toPath(to(d))));
}

export function drawCurve(selection: D3Selection, selector: string | undefined,
    {klass, from, to, inset, style, fn}:
    {klass?: string, from: NodePointFunction, to: NodePointFunction, inset: number, style?: Style, fn?: Func}): void {
    (!selector ? selection.append('path') : selection.select(selector))
        .call(attrIfPresent, 'class', klass)
        .call(styleIfPresent, style)
        .call(fnIfPresent, fn)
        .attr('d', (d: FeatureNode) => {
            const _from = from(d), _to = to(d);
            if (_from.x - inset < _to.x)
                logger.warn(() => 'too much inset or wrong order of points');
            return toD(MOVE, toPath(_to),
                CURVE, toPath(_from.x - inset, _to.y), toPath(_from.x - inset, _from.y), toPath(_from));
        });
}

export function drawCircle(selection: D3Selection, selector: string | undefined,
    {klass, center, radius, style, fn}:
    {klass?: string, from: NodePointFunction, to: NodePointFunction, center: NodePointFunction,
        radius: number, style?: Style, fn?: Func}): void {
    (!selector ? selection.append('circle') : selection.select(selector))
        .call(attrIfPresent, 'class', klass)
        .call(attrIfPresent, 'r', radius)
        .call(styleIfPresent, style)
        .call(fnIfPresent, fn)
        .call((selection: D3Selection) => {
            if (center)
                selection.attr('transform', (d: FeatureNode) => `translate(${toPath(center(d))})`);
        });
}

function polarToCartesian({x, y}: Point, radius: number, degrees: number): Point {
    const radians = degrees * Math.PI / 180.0;
    return {
        x: x + (radius * Math.cos(radians)),
        y: y + (radius * Math.sin(radians))
    };
}

export function cartesianToAngle(center: Point, point: Point): number {
    return Math.atan2(point.y - center.y, point.x - center.x) * 180.0 / Math.PI;
}

export function arcSegmentPath(center: Point, radius: number, startAngle: number, endAngle: number, sweepFlag = false): string {
    const start = polarToCartesian(center, radius, startAngle),
        end = polarToCartesian(center, radius, endAngle);
    return toD(MOVE, toPath(end), ARC, radius, radius, 0, 0, sweepFlag ? 1 : 0, toPath(start));
}

export function arcSlicePath(center: Point, radius: number, startAngle: number, endAngle: number, sweepFlag = false): string {
    return toD(arcSegmentPath(center, radius, startAngle, endAngle, sweepFlag), toD(LINE, toPath(center), CLOSE));
}

export function addStyle(selection: D3Selection, ...styleDescriptors: StyleDescriptor[]): void {
    styleDescriptors.forEach(styleDescriptor => {
        let {property, ...styles} = styleDescriptor;
        if (typeof property !== 'undefined')
            Object.keys(styles).forEach(key =>
                selection
                    .filter((node: FeatureNode) => node.feature().getPropertyString(property!) === key)
                    .call(selection => selection.attrs(styles[key])));
        else
            selection.attrs(styleDescriptor);
    });
}