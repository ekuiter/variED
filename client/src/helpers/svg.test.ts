import * as svg from './svg';
import {select as d3Select} from 'd3-selection';
import 'd3-selection-multi';
import {FeaturePropertyKey} from '../types';
import logger from './logger';

jest.mock('./logger');

const point = (x: number, y: number) => ({x, y}),
    pointFn = (x: number, y: number) => () => point(x, y);

describe('svg', () => {
    const createSvg = () => d3Select(document.createElement('svg'));

    describe('drawLine', () => {
        it('sets an attribute if present', () => {
            expect(createSvg().call(svg.attrIfPresent, 'attr', 'value').attr('attr')).toBe('value');
            expect(createSvg().call(svg.attrIfPresent, 'attr').attr('attr')).toBeNull();
        });

        it('sets a style if present', () => {
            expect(createSvg().call(svg.styleIfPresent, {width: 100}).attr('width')).toBe('100');
            expect(createSvg().call(svg.styleIfPresent).attr('width')).toBeNull();
        });

        it('calls a function if present', () => {
            const mock = jest.fn();
            createSvg().call(svg.fnIfPresent);
            expect(mock).not.toBeCalled();
            createSvg().call(svg.fnIfPresent, mock);
            expect(mock).toBeCalled();
        });

        it('appends a cross', () => {
            expect(createSvg().call(svg.appendCross)).toMatchSnapshot();
        });

        it('updates a rectangle', () => {
            expect(createSvg().call(svg.updateRect, {x: 10, y: 10, width: 100, height: 50})).toMatchSnapshot();
        });

        it('set a translation transformation', () => {
            expect(createSvg().call(svg.translateTransform, () => 10, () => 20).attr('transform'))
                .toBe('translate(10,20)');
        });

        it('draws a line', () => {
            expect(createSvg().call(svg.drawLine, null, {from: pointFn(0, 0), to: pointFn(10, 10)})).toMatchSnapshot();
        });
        
        it('updates a line', () => {
            const selection = createSvg();
            selection.append('path');
            selection.call(svg.drawLine, 'path', {from: pointFn(0, 0), to: pointFn(10, 10)});
            expect(selection).toMatchSnapshot();
        });

        it('draws a curve', () => {
            expect(createSvg().call(svg.drawCurve, null, {from: pointFn(10, 10), to: pointFn(0, 0), inset: 5})).toMatchSnapshot();
        });

        it('updates a curve', () => {
            const selection = createSvg();
            selection.append('path');
            selection.call(svg.drawCurve, 'path', {from: pointFn(10, 10), to: pointFn(0, 0), inset: 5});
            expect(selection).toMatchSnapshot();
        });

        it('warns when drawing a curve with invalid inset or point order', () => {
            (logger.warn as any).mockReset();
            const selection = createSvg();
            expect(logger.warn).not.toBeCalled();
            selection.call(svg.drawCurve, null, {from: pointFn(0, 0), to: pointFn(10, 10), inset: 5});
            expect(logger.warn).toBeCalled();
        });

        it('draws a circle', () => {
            expect(createSvg().call(svg.drawCircle, null, {radius: 5})).toMatchSnapshot();
        });

        it('draws a circle with a center', () => {
            expect(createSvg().call(svg.drawCircle, null, {center: pointFn(10, 10), radius: 5})).toMatchSnapshot();
        });

        it('updates a circle', () => {
            const selection = createSvg();
            selection.append('circle');
            selection.call(svg.drawCircle, 'circle', {center: pointFn(10, 10), radius: 5});
            expect(selection).toMatchSnapshot();
        });

        it('extracts an angle from cartesian coordinates', () => {
            expect(svg.cartesianToAngle(point(0, 0), point(0, 0))).toBe(0);
            expect(svg.cartesianToAngle(point(0, 0), point(10, 10))).toBe(45);
            expect(svg.cartesianToAngle(point(10, 10), point(20, 10))).toBe(0);
            expect(svg.cartesianToAngle(point(10, 10), point(10, 20))).toBe(90);
            expect(svg.cartesianToAngle(point(10, 10), point(-10, 10))).toBe(180);
            expect(svg.cartesianToAngle(point(10, 10), point(10, -10))).toBe(-90);
            expect(svg.cartesianToAngle(point(10, 10), point(-10, -10))).toBe(-135);
        });

        it('creates an arc segment path', () => {
            expect(svg.arcSegmentPath(point(10, 10), 5, 0, 90)).toBe('M 10,15 A 5 5 0 0 0 15,10');
        });

        it('creates an arc segment path with enabled sweep flag', () => {
            expect(svg.arcSegmentPath(point(10, 10), 5, 0, 90, true)).toBe('M 10,15 A 5 5 0 0 1 15,10');
        });

        it('creates an arc slice path', () => {
            expect(svg.arcSlicePath(point(10, 10), 5, 0, 90)).toBe('M 10,15 A 5 5 0 0 0 15,10 L 10,10 Z');
        });

        it('adds styles', () => {
            const selection = createSvg();
            selection.append('g');
            const g = selection.selectAll('g').data([
                {
                    feature: () => ({
                        getPropertyString: (key: FeaturePropertyKey) => {
                            if (typeof key === 'function')
                                return key(undefined!);
                            return key === 'yesProperty' ? 'yes' : 'no'
                        }
                    })
                }
            ]);
            g.call(svg.addStyle, {width: 100}, {
                property: 'yesProperty',
                yes: {shouldBeYes: 'yes'},
                no: {shouldBeYes: 'no'}
            }, {
                property: 'noProperty',
                yes: {shouldBeNo: 'yes'},
                no: {shouldBeNo: 'no'}
            }, {
                property: () => 'yes',
                yes: {shouldAlsoBeYes: 'yes'},
                no: {shouldAlsoBeYes: 'no'}
            });
            expect(g.attr('shouldBeYes')).toBe('yes');
            expect(g.attr('shouldBeNo')).toBe('no');
            expect(g.attr('shouldAlsoBeYes')).toBe('yes');
            expect(selection.node()).toMatchSnapshot();
        });
    });
});