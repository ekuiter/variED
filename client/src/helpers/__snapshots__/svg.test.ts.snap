// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svg drawLine adds styles 1`] = `
<svg>
  <g
    shouldalsobeyes="yes"
    shouldbeno="no"
    shouldbeyes="yes"
    width="100"
  />
</svg>
`;

exports[`svg drawLine appends a cross 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <path
          d="M -4 -4 L 4 4"
          fill="none"
          stroke="red"
          stroke-width="1.5px"
        />
        <path
          d="M -4 4 L 4 -4"
          fill="none"
          stroke="red"
          stroke-width="1.5px"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine draws a circle 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <circle
          r="5"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine draws a circle with a center 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <circle
          r="5"
          transform="translate(10,10)"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine draws a curve 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <path
          d="M 0,0 C 5,0 5,10 10,10"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine draws a line 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <path
          d="M 0,0 L 10,10"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine updates a circle 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <circle
          r="5"
          transform="translate(10,10)"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine updates a curve 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <path
          d="M 0,0 C 5,0 5,10 10,10"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine updates a line 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg>
        <path
          d="M 0,0 L 10,10"
        />
      </svg>,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;

exports[`svg drawLine updates a rectangle 1`] = `
Selection {
  "_groups": Array [
    Array [
      <svg
        fill="none"
        height="50"
        stroke="red"
        stroke-dasharray="4,4"
        width="100"
        x="10"
        y="10"
      />,
    ],
  ],
  "_parents": Array [
    null,
  ],
}
`;
